File: c++/125 - Valid Palindrome.cpp
// 125
// Valid Palindrome
// https://leetcode.com/problems/valid-palindrome/
// c++
// easy
// O(n)
// O(1)
// string

// time O(n)
// space O(1)
// first solution, not elegant but its fast and decent on memory
class Solution {
public:
    bool isPalindrome(string s) {
        int left = 0;
        int right = s.length()-1;
        
        while (left < right) {
            char lc = s.at(left);
            if (!isalnum(lc)) {
                left++;
                continue;
            }
            char rc = s.at(right);
            if (!isalnum(rc)) {
                right--;
                continue;
            }
            
            if (tolower(rc) != tolower(lc))
                return false;
            
            left++;
            right--;
        }
        
        return true;
    }
};

File: python/125 - Valid Palindrome.py
# 125
# Valid Palindrome
# https://leetcode.com/problems/valid-palindrome/
# python
# easy
# O(n)
# O(1)
# string


# time O(n)
# space O(n)
# python makes this an easy, elegant solution
# however, for a better solution with even faster results
# and more importantly, a space complexity of O(1)
# a index/pointer algorithim should be used
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = re.sub(r'[^a-z0-9]', '', s.lower())
        if s == s[::-1]:
            return True
        return False
